File: java/src/main/java/dev/intermediatebox/hashtable/RandomPickIndex.java
package dev.intermediatebox.hashtable;

import java.util.*;

public class RandomPickIndex {
  public static void main(String[] args) {
    int[] nums = new int[]{1, 2, 3, 3, 3};

    Solution1 s1 = new Solution1(nums);
    System.out.println(s1.pick(3));
    System.out.println(s1.pick(1));
    System.out.println(s1.pick(3));

    Solution2 s2 = new Solution2(nums);
    System.out.println(s2.pick(3));
    System.out.println(s2.pick(1));
    System.out.println(s2.pick(3));
  }

  // Tag: HashTable
  // Tag: List

  // O(n) time complexity, because of the list creation
  // O(n) space complexity for the list
  private static class Solution1 {
    private int[] nums;

    public Solution1(int[] nums) {
      this.nums = nums;
    }

    public int pick(int target) {
      List<Integer> list = new ArrayList<>();

      for (int i = 0; i < nums.length; i++) {
        if (nums[i] == target) {
          list.add(i);
        }
      }

      return list.get(new Random().nextInt(list.size()));
    }
  }

  // O(n) time complexity, because of the map creation; O(1) time complexity for pick operation
  // O(n) space complexity for the map
  private static class Solution2 {
    private int[] nums;

    private Map<Integer, ArrayList<Integer>> map;

    public Solution2(int[] nums) {
      this.nums = nums;

      this.map = new HashMap<>();

      for (int i = 0; i < nums.length; i++) {
        if (!map.containsKey(nums[i])) {
          map.put(nums[i], new ArrayList<>());
        }

        map.get(nums[i]).add(i);
      }
    }

    public int pick(int target) {
      List<Integer> indexes = map.get(target);
      return map.get(target).get(new Random().nextInt(indexes.size()));
    }
  }
}

File: typescript/src/hashtable/RandomPickIndex.ts
(() => {
  class Solution1 {
    private nums: number[];

    constructor(nums: number[]) {
      this.nums = nums;
    }

    pick(target: number): number {
      const indexes: number[] = [];

      for (let i = 0; i < this.nums.length; i++) {
        if (nums[i] === target) {
          indexes.push(nums[i]);
        }
      }

      return indexes[Math.floor(Math.random() * indexes.length)];
    }
  }

  class Solution2 {
    private map: Map<number, Array<number>>;
    private nums: number[];

    constructor(nums: number[]) {
      this.nums = nums;
      this.map = new Map();

      for (let i = 0; i < nums.length; i++) {
        if (!this.map.has(nums[i])) {
          this.map.set(nums[i], new Array());
        }

        this.map.get(nums[i])!.push(i);
      }
    }

    pick(target: number): number {
      return this.map.get(target)![Math.floor(Math.random() * this.map.get(target)!.length)];
    }
  }

  const nums = [1, 2, 3, 3, 3];


  const s1 = new Solution1(nums);
  console.log(s1.pick(3));
  console.log(s1.pick(1));
  console.log(s1.pick(3));

  const s2 = new Solution2(nums);
  console.log(s2.pick(3));
  console.log(s2.pick(1));
  console.log(s2.pick(3));
})();